### PR TITLE
Feature/reeanable april tags

### PR DESCRIPTION
**Pull request by DrCodeWizzard**

### CODE REVIEW

File: apriltags_ros/src/apriltag_detector.cpp [hunk 1]
```diff
@@ -452,29 +493,71 @@ tf::Transform AprilTagDetector::getDepthImagePlaneTransform(const sensor_msgs::P
...
+  // generate a projection from rgb_camera_reference frame to image frame that is the P matrix in the rgb_info object
+
+  // actually convert the whole point cloud from depth frame to rgb frame
+
+  std::for_each(pointCloud->begin(), pointCloud->end(), [&depth_to_rgb_transform](pcl::PointXYZRGB & point) -> void {
+    tf::Vector3 pcl_point(point.x, point.y, point.z); pcl_point = depth_to_rgb_transform * pcl_point; point.x = pcl_point.x(); point.y = pcl_point.y(); point.z = pcl_point.z(); });
+
+  //depth_to_rgb_transform
+    //
+    Eigen::Matrix3f projection_mat;
```
Comment (dantwinkler) [inline apriltags_ros/src/apriltag_detector.cpp]: I think you can also do: Eigen::Matrix3f intrinsicMatrix(rgb_info->K, 3, 3);

Then you can do:
  
// https://github.com/6RiverSystems/apriltags_ros/blob/hotfix/simplify-transform/apriltags_ros/src/apriltag_detector.cpp#L471
pcl::transformPointCloud (*pointCloud, *pointCloud, depthToRgbTf); // from getTransform
pcl::transformPointCloud (*pointCloud, *pointCloud, intrinsicMatrix);

Can also combine the two transforms before the transformation so there is only one pass over the point cloud.

Comment (DrCodeWizzard) [reply apriltags_ros/src/apriltag_detector.cpp]: Ohh I like that intrinsicMAtrix part :-)

File: apriltags_ros/src/apriltag_detector.cpp [hunk 2]
```diff
@@ -132,7 +137,7 @@ AprilTagDetector::AprilTagDetector(ros::NodeHandle& nh, ros::NodeHandle& pnh) :
 }
 
 AprilTagDetector::~AprilTagDetector(){
-
+    ROS_WARN("HERE %d %s shutting down", __LINE__, __FILE__);
```
Comment (dantwinkler) [inline apriltags_ros/src/apriltag_detector.cpp]: Remove.

File: apriltags_ros/src/apriltag_detector.cpp [hunk 3]
```diff
@@ -148,17 +153,24 @@ double absoluteAngleDiff(double angleA, double angleB)
...
-  const sensor_msgs::CameraInfoConstPtr& rgb_cam_info,
-  const sensor_msgs::CameraInfoConstPtr& depth_cam_info) {
+  const sensor_msgs::CameraInfoConstPtr& rgb_cam_info, const sensor_msgs::CameraInfoConstPtr& depth_cam_info) {
+
+  pid_t tid = syscall(SYS_gettid);
+
   // Check for trigger / timing
   if (!enabled_) {
     ROS_DEBUG_THROTTLE(5.0, "April images received but not enabled.");
+    ROS_WARN("HERE %d %d %s April images received but not enabled.", (int) tid, __LINE__, __FILE__);
```
Comment (dantwinkler) [inline apriltags_ros/src/apriltag_detector.cpp]: Remove.

File: apriltags_ros/src/apriltag_detector.cpp [hunk 4]
```diff
@@ -24,7 +24,12 @@
 #include <eigen_conversions/eigen_msg.h>
 #include <tf_conversions/tf_eigen.h>
 #include <cmath>
-
+#include <Eigen/Core>
+#define _GNU_SOURCE
+#include <unistd.h>
+#include <sys/syscall.h>
```
Comment (dantwinkler) [inline apriltags_ros/src/apriltag_detector.cpp]: Clean up unnecessary includes.

File: apriltags_ros/src/apriltag_detector.cpp [hunk 5]
```diff
@@ -206,6 +219,9 @@ void AprilTagDetector::imageCb(const sensor_msgs::PointCloud2ConstPtr& cloud,
       cv_ptr->header.frame_id = sensor_frame_id_;
     }
 
+    ROS_WARN("HERE %d %d %s", (int) tid, __LINE__, __FILE__);
```
Comment (dantwinkler) [inline apriltags_ros/src/apriltag_detector.cpp]: Remove all HERE logs

File: apriltags_ros/src/apriltag_detector.cpp [hunk 6]
```diff
@@ -456,15 +495,17 @@ tf::Transform AprilTagDetector::getDepthImagePlaneTransform(const sensor_msgs::P
...
-      polygon[i].first = (polygon[i].first - rgb_info->K[2]) / rgb_info->K[0] * depth_info->K[0] + depth_info->K[2];
-      polygon[i].second = (polygon[i].second - rgb_info->K[5]) / rgb_info->K[4] * depth_info->K[4] + depth_info->K[5];
-    }
-  }
+
+  // cannot register points to depth frame... we don't have depth so we are projecting to an unknown location in depth
+  //if(!same_frame_id_)
+  //{
+  //  ROS_DEBUG_THROTTLE(5.0, "Register polygon points from rgb to depth frame");
+  //  for(size_t i = 0; i < 4; i++)
```
Comment (dantwinkler) [inline apriltags_ros/src/apriltag_detector.cpp]: Remove older unused code.

File: apriltags_ros/src/apriltag_detector.cpp [hunk 7]
```diff
@@ -473,25 +514,168 @@ tf::Transform AprilTagDetector::getDepthImagePlaneTransform(const sensor_msgs::P
...
+
+  //std::for_each(pointCloud->begin(), pointCloud->end(), [Bx,By, &point_cloud_in_rgb_frame](pcl::PointXYZRGB & point) -> void {
+  //      pcl::PointXYZRGB pcl_point(point.x+Bx, point.y+By, point.z);
+  //      point_cloud_in_rgb_frame.push_back(pcl_point);
+  //});
+  point_cloud_in_rgb_frame.header.frame_id = rgb_camera_frame_name;
+  //depth_to_rgb_transform
+    //
+    Eigen::Matrix3f projection_mat;
+    projection_mat << rgb_info->P[0], rgb_info->P[1], rgb_info->P[2], rgb_info->P[4], rgb_info->P[5], rgb_info->P[6], rgb_info->P[8], rgb_info->P[9], rgb_info->P[10];
```
Comment (dantwinkler) [inline apriltags_ros/src/apriltag_detector.cpp]: Would like to use the pinhole model class as it is more concise and clear.

File: apriltags_ros/src/apriltag_detector.cpp [hunk 8]
```diff
@@ -473,25 +514,168 @@ tf::Transform AprilTagDetector::getDepthImagePlaneTransform(const sensor_msgs::P
...
+    }
+
+    for (int i = 0; i != 12; ++i) {
+        ROS_WARN("HERE %d %s depth_info->p[%d] = %f", __LINE__, __FILE__, i, depth_info->P[i]);
+    }
+  ROS_DEBUG_THROTTLE(5.0, "about to analyze the cloud");
+  ROS_WARN("HERE %d %s about to analyze the cloud", __LINE__, __FILE__);
+
+
+    float minX = 10000;
```
Comment (dantwinkler) [inline apriltags_ros/src/apriltag_detector.cpp]: We should remove this once we are comfortable with the results.

File: apriltags_ros/src/apriltag_detector.cpp [hunk 9]
```diff
@@ -473,25 +514,168 @@ tf::Transform AprilTagDetector::getDepthImagePlaneTransform(const sensor_msgs::P
...
+  float maxXrgb = -10000;
+  float maxYrgb = -10000;
+
+  double fx;
+  double fy;
+  double cx;
+  double tx = 0;
+  double cy;
+  double ty = 0;
+  if (true) {
```
Comment (dantwinkler) [inline apriltags_ros/src/apriltag_detector.cpp]: I think you are correct that we should use the same logic as the other section of the april tag code but we should use the same boolean argument instead (see ^^^).

File: apriltags_ros/src/apriltag_detector.cpp [hunk 10]
```diff
@@ -473,25 +514,168 @@ tf::Transform AprilTagDetector::getDepthImagePlaneTransform(const sensor_msgs::P
...
+  }
+
+
+  for (int x = 0; x < point_cloud_in_rgb_frame.width; x++)
   {
-    for (int y = 0; y < pointCloud->height; y++)
+    for (int y = 0; y < point_cloud_in_rgb_frame.height; y++)
     {
-        int i = x + (pointCloud->width * y);
+        int i = x + (point_cloud_in_rgb_frame.width * y);
```
Comment (dantwinkler) [inline apriltags_ros/src/apriltag_detector.cpp]: I would take the loop from my branch as it is more concise.

File: apriltags_ros/src/apriltag_detector.cpp [hunk 11]
```diff
@@ -473,25 +514,168 @@ tf::Transform AprilTagDetector::getDepthImagePlaneTransform(const sensor_msgs::P
...
+    //}
+    pcl::PointXYZ pcl_point{u_rgb, v_rgb, 0};
+    ROS_WARN("HERE %d %s point %d with xyz coordinates %f %f %f with rgb coordinates %f %f %f has image projection %f %f %f is an inliner", __LINE__, __FILE__, i, pointCloud->points[i].x, pointCloud->points[i].y, pointCloud->points[i].z, point_cloud_in_rgb_frame.points[i].x, point_cloud_in_rgb_frame.points[i].y, point_cloud_in_rgb_frame.points[i].z, pcl_point.x, pcl_point.y, pcl_point.z);
+
+
+  }
 
   pcl::ExtractIndices<pcl::PointXYZRGB> extract;
-  extract.setInputCloud(pointCloud);
+  extract.setInputCloud(point_cloud_in_rgb_frame.makeShared());
```
Comment (dantwinkler) [inline apriltags_ros/src/apriltag_detector.cpp]: is makeShared necessary?

Comment (DrCodeWizzard) [reply apriltags_ros/src/apriltag_detector.cpp]: yes

**Review by dantwinkler — APPROVED** [review]

**Pull request merged**